### DISCUSSION TITLE
Trouble Deploying Wave apps with Uvicorn

### DISCUSSION THREAD

**Discussion by tolmanam** [h2oai/wave#1803] [Q&A] [answered]
I don't understand the interactions between the `waved` server and the `wave apps`.  I'd like to be able to spin up the `waved` server and then run my applications with `uvicorn` as described in the [Deploying Wave apps docs](https://wave.h2o.ai/docs/deployment#deploying-wave-apps) but the two commands provided **do not actually seem** to be equivalent.

```
(venv) $ wave run --no-reload --no-autostart foo
```

```
(venv) $ uvicorn foo:main
```

Both seem to spin up the application API, but when I start the app with the `wave` command it also connects up to the running `waved` service and registers the route.  The `uvicorn` command does not.

Am I missing a step?

Thanks.

_versions:_
 * _waved - 0.24.2 20221209122225_
 * _h2o-wave - 0.24.2_
 * _uvicorn - 0.20.0_

**Comment by mturoci** [comment, marked as answer]
I will have a proper answer for you on Monday, but if you would like to investigate on your own, you can check https://github.com/h2oai/wave/blob/94cb690ff3c55c9ea68bd29753e6b800fd97074e/py/h2o_wave/cli.py#L77 to get an idea of what wave run is actually doing.

I would also check if your config env variables  are set correctly, e.g. H2O_WAVE_ADDRESS (address of the your wave server).

Please paste any outputs/errors you get along the way - will speed up my investigation.

I agree the deployment docs need some love though. I plan to write detailed instructions for deployments on AWS EC2, Render etc. but didnt have time so far.

**Comment by mturoci** [comment]
Seems to work fine for me.

![image](https://user-images.githubusercontent.com/64769322/213990820-bb5fa3ca-e33b-4514-86a6-04baf72cb667.png)

@tolmanam Could you please paste screenshots of how you run both waved and uvicorn + the outputs you get? Do you set any wave env variables?

**Reply by tolmanam** [reply]
Ah!  I was spoiled by how easy it was to just run 'wave' and didn't realize that I needed to provide the 
_H2O_WAVE_APP_ADDRESS_ environment variable.

The following command seems to "work".

```
> H2O_WAVE_APP_ADDRESS="http://127.0.0.1:8000" uvicorn counter:main
```

Would you consider that a reasonable way to set that variable?  I'm happy to submit a merge request to update that documentation.

**Reply by mturoci** [reply]
>Would you consider that a reasonable way to set that variable?

Yes, this looks good.

> I needed to provide the H2O_WAVE_APP_ADDRESS environment variable

Ah, I see. `uvicorn app:main` works as long as default ports are used (http://localhost:10101 for wave server and http://localhost:8000 for wave app). If any of those is changed, it crashes. Would be great to make this clear in the docs for the future if you would like.
